ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/PackageValidator.scala

```diff
@@ -0,0 +1,154 @@
+        case Right(newPkgs) =>
+          StepContinue((uploadEntry, Some(newPkgs)))
+        case Left(errors) =>
+          val msg = errors.mkString(", ")
+          rejectionTraceLog(msg, uploadEntry)
+          reject(
+            ctx.getRecordTime,
+            uploadEntry.getSubmissionId,
+            uploadEntry.getParticipantId,
+            _.setInvalidPackage(DamlKvutils.Invalid.newBuilder.setDetails(msg))
+          )
+      }
+  }
+
+  // validate package that have been decoded in a previous step.
+  private[this] val validateEntry: Step = {
+    case (ctx, partialResult @ (uploadEntry, Some(pkgs))) =>
+      type Result = Either[List[String], Unit]
+
+      val result: Result = metrics.daml.kvutils.committer.packageUpload.validateTimer.time { () =>
+        val allPkgIds = uploadEntry.getArchivesList
+          .iterator()
+          .asScala
+          .map(pkg => Ref.PackageId.assertFromString(pkg.getHash))
+          .toSet
+
+        pkgs.keys.foldLeft[Result](Right(()))(
+          (acc, pkgId) =>
+            engine
+              .validatePackages(allPkgIds, pkgs)
+              .fold(
+                err => Left(s"Cannot decode archive ${pkgId}: ${err}" :: acc.left.getOrElse(Nil)),
+                _ => acc
+            )
+        )
+      }
+
+      result match {
+        case Right(_) =>
+          StepContinue(partialResult)
+        case Left(errors) =>
+          val msg = errors.mkString(", ")
+          rejectionTraceLog(msg, uploadEntry)
+          reject(
+            ctx.getRecordTime,
+            uploadEntry.getSubmissionId,
+            uploadEntry.getParticipantId,
+            _.setInvalidPackage(DamlKvutils.Invalid.newBuilder.setDetails(msg))
+          )
+      }
+  }
+
+  private val preload: Step = {
+    case (ctx, partialResult @ (uploadEntry, Some(pkgs))) =>
+      val errors: List[String] =
+        metrics.daml.kvutils.committer.packageUpload.preloadTimer.time { () =>
+          pkgs.iterator.flatMap {
+            case (pkgId, pkg) =>
+              engine
+                .preloadPackage(pkgId, pkg)
+                .consume(_ => None, pkgs.get, _ => None)
+                .fold(err => List(err.detailMsg), _ => List.empty)
+          }.toList
+        }
+
+      if (errors.isEmpty)
+        StepContinue(partialResult)
+      else {
+        val msg = errors.mkString(", ")
+        rejectionTraceLog(msg, uploadEntry)
+        reject(
+          ctx.getRecordTime,
+          uploadEntry.getSubmissionId,
+          uploadEntry.getParticipantId,
+          _.setInvalidPackage(DamlKvutils.Invalid.newBuilder.setDetails(msg))
+        )
+      }
+  }
+
+  override protected val steps: Iterable[(StepInfo, Step)] =
+    List(
+      "authorize_submission" -> authorizeSubmission,
+      "deduplicate_submission" -> deduplicateSubmission,
+      "decode_entry" -> decodeEntry,
+      "validate_entry" -> validateEntry,
+      "filter_duplicates" -> filterDuplicates,
+      "preload" -> preload,
```

Comment:
Do we need synchronous preloading ?

Reply:
No I don't think we need that, unless it's more efficient to validate+preload when committing rather than validating and loading the package to engine separately.

Reply:
Actually, I think the system is more predictable if we preload it synchronously when committing as that means as soon as the package upload completes we can start using the package without having a surprising delay on the first transaction. What do you think?

Reply:
I see a trade off: 
* If we preload we are indeed more predictable. It is also slightly more efficient as we decode only once. (I think our bigger dars need around 20s to decode).
* If we don't, we do not overload the engine with useless packages. Potentially the code for scenario/trigger/script can end in the dar even it is useless for the Ledger engine.

Reply:
I see the feature useful for testing/development. In production, dar upload should be so seldom that it should not really matter.

Reply:
I think the preload should by synchronous for the pro-preload reasons you both mentioned.

Reply:
We will preload in the 3 cases.